File: Go/src/0011 - Container With Most Water.go
// 链接：https://leetcode.com/problems/container-with-most-water/
// 题意：给定 n 个非负整数 a1...an，表示有端点分别在 (i, 0), (i, ai) 的线段，求任意两条线段组成的矩形的最大面积

// 输入：[1,8,6,2,5,4,8,3,7]
// 输出：49

// 思路：双指针，维护两个指针 i, j，分别初始化为 0，n - 1，则其面积为：area = (j - i) * min(height[i], height[j])
//		(1) 如果 height[i] <= height[j]，则只能向右移动 i
//			假设向左移动 j，则面积必定变小，因为：
//			   (j - 1 - i) * min(height[i], height[j - 1])
//			<= (j - 1 - i) * height[i]
//			<  (j - i) * height[i] = area
//		(2) 如果 height[i] >  height[j]，则只能向左移动 j
//			假设向右移动 i，则面积必定变小，因为：
//			   (j - (i + 1)) * min(height[i + 1], height[j])
//			<= (j - (i + 1)) * height[j]
//			<  (j - i) * height[j] = area
//		时间复杂度：O(n)，空间复杂度：O(1)

// 这题求的结果和另一题（https://leetcode.com/problems/largest-rectangle-in-histogram/）类似
// 刚开始一直以为条件都一样，就用了单调栈，结果样例就跑不过（其实想的时候在大脑中跑就感觉不对，有很多反例）
// 后来仔细一对比，结果发现本题的结果只与选择的两条线段有关
// 而另一题是要求连续的高度都大于等于矩形的高度才行
// 还是需要仔细审题，不能凭借经验胡乱猜测题意
func maxArea(height []int) int {
	l, r := 0, len(height) - 1
	result := 0
	for ; l < r;  {
		result = max(result, (r - l) * min(height[l], height[r])) // 更新面积最大值
		if height[l] <= height[r] {
			l += 1
		} else {
			r -= 1
		}
	}
	
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: Rust/src/0001 ~ 0100/0011 - Container With Most Water.rs
// 链接：https://leetcode.com/problems/container-with-most-water/
// 题意：给定一个长度为 n 的非负整数数组 height ，
//      其中 height[i] 表示在端点为 (i, 0) 和 (i, height[i]) 的线段，
//      求任意两条线段组成的矩形的最大面积？


// 数据限制：
//  n == height.length
//  2 <= n <= 10 ^ 5
//  0 <= height[i] <= 10 ^ 4


// 输入： height = [1,8,6,2,5,4,8,3,7]
// 输出： 49

// 输入： height = [1,1]
// 输出： 1


// 思路： 双指针
//
//      维护题目所需的最大面积 ans ，初始化为 0 。
//
//      我们定义左右指针 l 和 r ，分别初始化为 0 和 n - 1 。
//
//      每次计算当前左右指针对应的线段组成的矩形的面积，
//      即 area = min(height[l], height[r]) * (r - l) 。
//
//      然后更新 ans 的最大值，即 ans = max(ans, area) 。
//
//      接下来按照以下规则更新左右指针，并重复上述过程，直至 l >= r ：
//          1. height[l] <= height[r]: l 向右移动一位，r 不动。
//			   假设 r 向左移动一位，则面积必定变小，
//              因为： (r - 1 - l) * min(height[l], height[r - 1])
//			       <= (r - 1 - l) * height[l]
//			       <  (r - l) * height[i] = area
//          2. height[l] > height[r]: r 向左移动一位，l 不动。
//			   假设 l 向右移动一位，则面积必定变小，
//              因为： (r - (l + 1)) * min(height[l + 1], height[r])
//			       <= (r - (l + 1)) * height[r]
//			       <  (r - l) * height[r] = area
//
//
//      时间复杂度：O(n)
//          1. 需要遍历 height 中全部 O(n) 个数字
//      空间复杂度：O(1)
//          1. 只需要维护常数个额外变量


impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        // 维护题目所需的最大面积 ans ，初始化为 0
        let mut ans = 0;
        // 左指针 l ，初始化为 0
        let mut l = 0;
        // 右指针 r ，初始化为 height.len() - 1
        let mut r = height.len() - 1;
        // 当矩形底还存在时，继续循环
        while l < r {
            // 计算当前矩形的面积
            let area = (r - l) as i32 * (height[l].min(height[r]));
            // 更新 ans 的最大值
            ans = ans.max(area);
            if height[l] <= height[r] {
                // 如果 height[l] <= height[r] ，
                // 则 l 向右移动一位，r 不动
                l += 1;
            } else {
                // 如果 height[l] > height[r] ，
                // 则 r 向左移动一位，l 不动
                r -= 1;
            }
        }

        ans
    }
}
